File: app/src/main/java/org/dhis2/utils/ValueUtils.java
package org.dhis2.utils;

import org.hisp.dhis.android.core.D2;
import org.hisp.dhis.android.core.common.ValueType;
import org.hisp.dhis.android.core.option.Option;
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue;

import java.util.Objects;

/**
 * QUADRAM. Created by ppajuelo on 25/09/2018.
 */

public class ValueUtils {

    private ValueUtils() {
        throw new IllegalStateException("Utility class");
    }

    public static TrackedEntityAttributeValue transform(D2 d2, TrackedEntityAttributeValue attributeValue, ValueType valueType, String optionSetUid) {
        String transformedValue = transformValue(d2, attributeValue.value(), valueType, optionSetUid);

        if (!Objects.equals(transformedValue, attributeValue.value())) {
            return attributeValue.toBuilder()
                    .value(transformedValue)
                    .build();
        } else {
            return attributeValue;
        }
    }

    public static String transformValue(D2 d2, String value, ValueType valueType, String optionSetUid) {
        String teAttrValue = value;
        if (valueType.equals(ValueType.ORGANISATION_UNIT)) {
            if (!d2.organisationUnitModule().organisationUnits().byUid().eq(value).blockingIsEmpty()) {
                String orgUnitName = d2.organisationUnitModule().organisationUnits()
                        .byUid().eq(value)
                        .one().blockingGet().displayName();
                teAttrValue = orgUnitName;
            }
        } else if (optionSetUid != null) {
            String optionCode = value;
            if (optionCode != null) {
                Option option = d2.optionModule().options().byOptionSetUid().eq(optionSetUid).byCode().eq(optionCode).one().blockingGet();
                if (option != null && (Objects.equals(option.code(), optionCode) || Objects.equals(option.name(), optionCode))) {
                    teAttrValue = option.displayName();
                }
            }
        }
        return teAttrValue;
    }
}

File: app/src/main/java/org/dhis2/utils/ValueUtils.kt
package org.dhis2.utils

import org.dhis2.commons.extensions.toFriendlyDate
import org.dhis2.commons.extensions.toFriendlyDateTime
import org.dhis2.commons.extensions.toPercentage
import org.hisp.dhis.android.core.D2
import org.hisp.dhis.android.core.common.ValueType
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue

/**
 * QUADRAM. Created by ppajuelo on 25/09/2018.
 */
class ValueUtils private constructor() {
    init {
        throw IllegalStateException("Utility class")
    }

    companion object {
        fun transform(
            d2: D2,
            attributeValue: TrackedEntityAttributeValue,
            valueType: ValueType?,
            optionSetUid: String?,
        ): TrackedEntityAttributeValue {
            val transformedValue =
                transformValue(d2, attributeValue.value(), valueType, optionSetUid)

            return if (transformedValue != attributeValue.value()) {
                attributeValue.toBuilder()
                    .value(transformedValue)
                    .build()
            } else {
                attributeValue
            }
        }

        fun transformValue(
            d2: D2,
            value: String?,
            valueType: ValueType?,
            optionSetUid: String?,
        ): String? {
            var teAttrValue = value
            when (valueType) {
                ValueType.ORGANISATION_UNIT -> {
                    if (!d2.organisationUnitModule().organisationUnits().byUid().eq(value)
                            .blockingIsEmpty()
                    ) {
                        val orgUnitName = d2.organisationUnitModule().organisationUnits()
                            .byUid().eq(value)
                            .one().blockingGet()!!.displayName()!!
                        teAttrValue = orgUnitName
                    }
                }
                ValueType.DATE, ValueType.AGE -> {
                    teAttrValue = teAttrValue?.toFriendlyDate()
                }
                ValueType.DATETIME -> {
                    teAttrValue = teAttrValue?.toFriendlyDateTime()
                }

                ValueType.PERCENTAGE -> {
                    teAttrValue = teAttrValue?.toPercentage()
                }
                else -> {
                    teAttrValue = transformOptionSet(optionSetUid, d2, value)
                }
            }
            return teAttrValue
        }

        private fun transformOptionSet(optionSetUid: String?, d2: D2, value: String?): String? {
            var teAttrValue = value
            if (optionSetUid != null) {
                val optionCode = value
                if (optionCode != null) {
                    val option =
                        d2.optionModule().options().byOptionSetUid().eq(optionSetUid).byCode()
                            .eq(optionCode).one().blockingGet()
                    if (option != null && (option.code() == optionCode || option.name() == optionCode)) {
                        teAttrValue = option.displayName()
                    }
                }
            }
            return teAttrValue
        }
    }
}
